src/katgpucbf/xbgpu/beamform.py

```diff
@@ -0,0 +1,151 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Implement the calculations for beamforming."""
+
+from importlib import resources
+from typing import Sequence
+
+import numpy as np
+from katsdpsigproc import accel
+from katsdpsigproc.abc import AbstractCommandQueue, AbstractContext
+
+from .. import COMPLEX, N_POLS
+
+
+class BeamformTemplate:
+    """Template for beamforming.
+
+    Parameters
+    ----------
+    context
+        The GPU context that we'll operate in.
+    pols
+        One entry per single-polarisation output beam. Each entry is either
+        0 or 1, to indicate which input polarisation to use in the beam.
+    """
+
+    def __init__(self, context: AbstractContext, beam_pols: Sequence[int]) -> None:
+        # TODO: tune these. And maybe adapt to input shape?
+        self.block_time = 128
+        self.block_channels = 1
+        self.beam_pols = beam_pols
+        with resources.as_file(resources.files(__package__)) as resource_dir:
+            program = accel.build(
+                context,
+                "kernels/beamform.mako",
+                {
+                    "block_time": self.block_time,
+                    "block_channels": self.block_channels,
+                    "beam_pols": self.beam_pols,
+                },
+                extra_dirs=[str(resource_dir)],
+            )
+        self.kernel = program.get_kernel("beamform")
+
+    def instantiate(
+        self,
+        command_queue: AbstractCommandQueue,
+        frames: int,
+        antennas: int,
+        channels: int,
+        times: int,
+    ) -> "Beamform":
+        """Generate a :class:`Beamform` object based on the template."""
+        return Beamform(self, command_queue, frames, antennas, channels, times)
+
+
+class Beamform(accel.Operation):
+    r"""Operation for beamforming.
+
+    .. rubric:: Slots
+
+    **in** : frames × antennas × channels × times × N_POLS × COMPLEX, int8
+        Complex (Gaussian integer) input channelised voltages
+    **out** : frames × beams × channels × times × COMPLEX, int8
+        Complex (Gaussian integer) output channelised voltages
+    **weights** : antennas × beams, complex64
+        Complex scale factor to apply to each antenna for each beam
+    **delays** : antennas × beams, float32
+        Delay used to compute channel-dependent phase rotation. The
+        rotation applied is :math:`e^{\pi cd}` where :math:`c` is
+        the channel number and :math:`d` is the delay value. Note
+        that this will not apply any rotation to the first channel
+        in the data; any such rotation needs to be baked into **weights**.
```

Comment:
> Note that this will not apply any rotation to the first channel in the data; any such rotation needs to be backed in **weights**.

How would/does that work?

Reply:
Multiply all the weights by exp(j * phase).